cds_books/literatures/tasks.py

```diff
@@ -0,0 +1,124 @@
+
+logger = logging.getLogger('tasks')
+
+MIN_CONTENT_LENGTH = 128
+
+
+def is_allowed(record):
+    """Allow only documents and series."""
+    schema = record.get("$schema")
+    if (
+        schema.endswith("document-v1.0.0.json") or
+        schema.endswith("series-v1.0.0.json")
+    ):
+        return True
+    return False
+
+
+def update_cover(sender, *args, **kwargs):
+    """Update cover metadata of the record with identifier."""
+    record = kwargs.get("record", {})
+    if not is_allowed(record):
+        return
+
+    cover_meta = record.get("cover_metadata", {})
+    if cover_meta.get("isbn"):
+        return
+
+    identifiers = record.get("identifiers")
+    if not identifiers:
+        return
+
+    if "cover_metadata" not in record:
+        record["cover_metadata"] = {}
+
+    schema = record.get("$schema")
+    if schema.endswith("series-v1.0.0.json"):
+        for ident in identifiers:
+            if ident["scheme"] != "ISSN":
+                continue
+            record["cover_metadata"]["issn"] = ident["value"]
+            return
+
+    for ident in identifiers:
+        if ident["scheme"] != "ISBN":
+            continue
+        record["cover_metadata"]["isbn"] = ident["value"]
+        return
+
+
+def select_cover(sender, *args, **kwargs):
+    """Triggers async task to set cover metadata of the record."""
+    record = kwargs.get("record", {})
+    if is_allowed(record):
+        select_cover_task.apply_async((record,))
+
+
+def is_cover_valid(record):
+    """Fetch a cover from url and returns only if it is valid."""
+    url = build_literatures_cover_urls(record)['small']
+    try:
+        resp = urllib.request.urlopen(url)
+        is_success = resp.getcode() == 200
+        more_than_1_pixel = resp.get("Content-Length") > MIN_CONTENT_LENGTH
+    except Exception as ex:
+        return None
+
+    if is_success and more_than_1_pixel:
+        return resp
+    return None
+
+
+@shared_task(ignore_result=True)
+def select_cover_task(record):
+    """Set record cover metadata isbn with valid cover."""
+    if not is_allowed(record):
+        return
+
+    if is_cover_valid(record):
+        return
+
+    identifiers = record.get("identifiers", [])
+    if not identifiers:
+        return
+
+    if "cover_metadata" not in record:
+        record["cover_metadata"] = {}
+
+    schema = record.get("$schema")
+    if schema.endswith("series-v1.0.0.json"):
+        for ident in identifiers:
+            if ident["scheme"] != "ISSN":
+                continue
+
+            record["cover_metadata"]["issn"] = ident["value"]
+            if is_cover_valid(record):
+                record.commit()
+                db.session.commit()
+                return
+
+    for ident in identifiers:
+        if ident["scheme"] != "ISBN":
+            continue
+
+        record["cover_metadata"]["isbn"] = ident["value"]
+        if is_cover_valid(record):
+            record.commit()
+            db.session.commit()
+            return
```

Comment:
there might be an issue here: if I have many ISSNs, but none of them are valid, and one valid ISBN, what will happen is:
```
cover_metadata: {
    issn: 342423434  # last ISSN that was not valid
    isbn:  342344342   # valid ISBN
}
```
I would avoid changing the record metadata if not needed.